### PR TITLE
[KeyConnector] Add support for key connector OTP

### PR DESCRIPTION
**Pull request by jlf0dev**
## Type of change
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

Add support for Key Connector in mobile

## Code changes

Please refer to bitwarden/jslib#527 and bitwarden/web#1256 for similar implementations.

## Testing requirements

- Server backwards compatibility needs testing (older version of server). Mainly logging in with both MP and SSO.

## Before you submit
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

**Comment by Hinton** [comment]
Left to implement:
* Complete ConvertMasterPasswordPage
* Add TOTP support for vault export

File: src/Core/Models/Request/TokenRequest.cs [hunk 1]
```diff
@@ -19,9 +19,10 @@ public class TokenRequest
         public bool? Remember { get; set; }
         public string CaptchaToken { get; set; }
         public DeviceRequest Device { get; set; }
+        public string OrgIdentifier { get; set; }
```
Comment (eliykat) [inline src/Core/Models/Request/TokenRequest.cs:22]: I think the changes in this file are based off some old commits on jslib, which I ended up reverting. You should be able to revert the changes in this file.

File: src/Core/Services/AuthService.cs [hunk 2]
```diff
@@ -275,7 +278,7 @@ private async Task<SymmetricCryptoKey> MakePreloginKeyAsync(string masterPasswor
         private async Task<AuthResult> LogInHelperAsync(string email, string hashedPassword, string localHashedPassword,
             string code, string codeVerifier, string redirectUrl, SymmetricCryptoKey key,
             TwoFactorProviderType? twoFactorProvider = null, string twoFactorToken = null, bool? remember = null,
-            string captchaToken = null)
+            string captchaToken = null, string orgId = null)
```
Comment (eliykat) [inline src/Core/Services/AuthService.cs:281]: Same as the changes to `TokenRequest`, some (but not all) of these changes were later reverted. Please compare to the current diff on the jslib PR.

File: src/Core/Services/AuthService.cs [hunk 3]
```diff
@@ -353,27 +356,85 @@ private async Task<SymmetricCryptoKey> MakePreloginKeyAsync(string masterPasswor
...
+
+                if (code == null || tokenResponse.Key != null)
+                {
+                    if (tokenResponse.KeyConnectorUrl != null)
                     {
-                        var keyPair = await _cryptoService.MakeKeyPairAsync();
-                        await _apiService.PostAccountKeysAsync(new KeysRequest
+                        try
+                        {
+                            var userKeyResponse = await _apiService.GetUserKeyFromKeyConnector(tokenResponse.KeyConnectorUrl);
```
Comment (eliykat) [inline src/Core/Services/AuthService.cs:375]: This should use `KeyConnectorService.getAndSetKey()`, which basically just extracts this code block to a service.

**Review by eliykat — COMMENTED** [review]
Thanks for your help with this @jlf0dev. I've left some comments below, although I know it's still a WIP so you might already be aware of some of this.

I'd also like @Hinton to review this at some stage, maybe when it's out of draft.

**Comment by eliykat** [comment]
I've added:
* hide "Change Master Password" if using key connector
* OTP verification on export component

A couple of notes:
* In jslib, we extracted the master password/OTP verification template into `VerifyMasterPasswordComponent`. That's a bit beyond my mobile skills so I've just put it directly in the export page for now - but it would be nice to fix so that it reflects jslib.
* the export component doesn't look great, it would be nice to space out the layout a bit now that there's the extra button in there:
Screen Shot 2021-11-09 at 9 45 05 pm

File: src/Core/Models/Request/TokenRequest.cs (hunk 1 above)
Comment (jlf0dev) [reply src/Core/Models/Request/TokenRequest.cs]: 84c25e0760ba3fedc0e235ca48a330253f96ef81

File: src/App/Resources/AppResources.resx [hunk 4]
```diff
@@ -2027,6 +2027,18 @@
   <data name="UpdatePasswordError" xml:space="preserve">
     <value>Currently unable to update password</value>
   </data>
+  <data name="RemoveMasterPassword" xml:space="preserve">
+    <value>Remove Master Password</value>
+  </data>
+  <data name="RemoveMasterPasswordWarning" xml:space="preserve">
+    <value>{0} is using SSO with customer-managed encryption. Continuing will remove your Master Password from your account and require SSO to login.&#10;&#10;If you do not want to remove your Master Password, you may leave this organization.</value>
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Resources/AppResources.resx]: 901dc76f61f185616f6f51781a06cfabe3ca403c

File: src/App/Pages/Accounts/RemoveMasterPasswordPage.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                    AppResources.Yes, AppResources.No);
+                if (confirm)
+                {
+                    await _vm.LeaveOrganization();
+                    await Navigation.PopModalAsync();
+                }
+            }
+        }
+
+        //private async void Close_Clicked(object sender, System.EventArgs e)
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/Accounts/RemoveMasterPasswordPage.xaml.cs]: 901dc76f61f185616f6f51781a06cfabe3ca403c

File: src/App/Pages/Accounts/RemoveMasterPasswordPage.xaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    xmlns:pages="clr-namespace:Bit.App.Pages"
+    xmlns:u="clr-namespace:Bit.App.Utilities"
+    x:DataType="pages:RemoveMasterPasswordPageViewModel"
+    Title="{Binding PageTitle}">
+
+    <ContentPage.BindingContext>
+        <pages:RemoveMasterPasswordPageViewModel />
+    </ContentPage.BindingContext>
+
+    <!--<ContentPage.ToolbarItems>
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/Accounts/RemoveMasterPasswordPage.xaml]: 901dc76f61f185616f6f51781a06cfabe3ca403c

File: src/Core/Services/AuthService.cs (hunk 3 above)
Comment (jlf0dev) [reply src/Core/Services/AuthService.cs]: 901dc76f61f185616f6f51781a06cfabe3ca403c

File: src/Core/Services/KeyConnectorService.cs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            _cryptoService = cryptoService;
+            _storageService = storageService;
+            _tokenService = tokenService;
+            _apiService = apiService;
+
+        }
+
+        public Task GetAndSetKey(string url = null)
+        {
+            //if (url == null)
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/Core/Services/KeyConnectorService.cs]: 901dc76f61f185616f6f51781a06cfabe3ca403c

File: src/App/Pages/TabsPage.cs [hunk 8]
```diff
@@ -70,6 +72,15 @@ public TabsPage(AppOptions appOptions = null, PreviousPageInfo previousPage = nu
             {
                 ResetToSendPage();
             }
+            
+        }
+
+        protected override async void OnAppearing()
+        {
+            if (await _keyConnectorService.UserNeedsMigration())
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/TabsPage.cs]: I looked into this pretty extensively today. I thought we could move the page to `App.xaml.cs` and use the MessagingService to bring it up after a sync. The problem is that we don't sync when we land in the vault. We sync when we authenticate with SSO or login, but if the vault is just locked there is no sync. 

I also considered saving "needs migration" to storage on the sync, then checking later and calling `RemoveMasterPasswordPage`. This seemed just a roundabout way of calling the page though and I didn't see an advantage. We would still have to attach it to a page somewhere.

I ended up just attaching it to the `LockPage`. 

Let me know what you think @eliykat, here's the commit:
af6525dcafb04eaea763d437593826ed1f9f9235

File: src/App/Services/MobilePasswordRepromptService.cs [hunk 9]
```diff
@@ -34,5 +40,11 @@ public async Task<bool> ShowPasswordPromptAsync()
 
             return await _platformUtilsService.ShowPasswordDialogAsync(AppResources.PasswordConfirmation, AppResources.PasswordConfirmationDesc, validator);
         }
+
+        public async Task<bool> Enabled()
+        {
+            var keyConnectorService = ServiceContainer.Resolve<IKeyConnectorService>("keyConnectorService");
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Services/MobilePasswordRepromptService.cs:46]: This is a little bit of a circular dependency problem. `MobilePasswordRepromptService` is actually created in platform specific code (for iOS its `iOSCoreHelpers`) before our `ServiceContainer` is created. The `KeyConnectorService` constructor uses many services created in the `ServiceContainer` so we can't register it sooner. But the `MobilePasswordRepromptService` uses the `KeyConnectorService`.

Anyways, thats all to say we can't put it in the constructor yet since it hasn't been created.

If you can think of a way around that @eliykat let me know.

File: src/Core/Services/KeyConnectorService.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                throw e;
+            }
+
+            await _apiService.PostConvertToKeyConnector();
+        }
+
+        public async Task<bool> UserNeedsMigration()
+        {
+            var loggedInUsingSso = _tokenService.GetIsExternal();
```
[the remark this answers is not in the document]
Comment (eliykat) [reply src/Core/Services/KeyConnectorService.cs:92]: Ignore my comment above, this can stay as it is.

**Comment by jlf0dev** [comment]
Left to implement:

- Backwards compatibility for models
- Hide "Master Password" Lock screen

File: src/Core/Models/Response/ProfileOrganizationResponse.cs [hunk 11]
```diff
@@ -25,5 +25,8 @@ public class ProfileOrganizationResponse
         public bool Enabled { get; set; }
         public Permissions Permissions { get; set; } = new Permissions();
         public string Identifier { get; set; }
+        public bool UsesKeyConnector { get; set; }
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/Core/Models/Response/ProfileOrganizationResponse.cs:28]: I'm actually not sure how we approach this yet. @mportune-bw could help me out here?

File: src/App/Pages/TabsPage.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply src/App/Pages/TabsPage.cs]: > The problem is that we don't sync when we land in the vault. We sync when we authenticate with SSO or login, but if the vault is just locked there is no sync.

But the client won't know if the user uses key connector _until_ there's a sync. A sync is the earliest possible time that we can know key connector is involved and redirect to the remove-password page. Your first suggestion of moving it to `App.xaml.cs` and using the messagingService seems fine to me if that was your only concern.

[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/TabsPage.cs]: I see your point. I believe we'll still have to attach the redirect to a page though. The initial sync happens before we reach the vault, while the SSO, lock, and login pages are showing. We can avoid showing it on those pages but unless the user initiates a sync manually, they won't be prompted with the remove-password page.

[the remark this answers is not in the document]
Comment (eliykat) [reply src/App/Pages/TabsPage.cs]: `TabsPage` is the common landing page when you login or unlock, right? That seems reasonable, unless @Hinton has a better idea.

File: src/App/Services/MobilePasswordRepromptService.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply src/App/Services/MobilePasswordRepromptService.cs:46]: Ok, that makes sense and that seems like a reasonable way to deal with it.

File: src/Core/Models/Response/ProfileOrganizationResponse.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply src/Core/Models/Response/ProfileOrganizationResponse.cs:28]: Looking at the other response models, it looks like we handle this by assigning default values. e.g. https://github.com/bitwarden/mobile/blob/90b62d61aef64a0ab84ab869d7682f3af8ff10b3/src/Core/Models/Response/CipherResponse.cs#L23

This is an educated guess though, please feel free to confirm.

**Review by eliykat — COMMENTED** [review]
Just a few minor comments. I've also made some style fixes.

Other than the few pain points we've discussed, all looks like a straightforward reimplementation, but please get Oscar to have a look (and I have code in here that needs to be reviewed).

File: src/Core/Services/AuthService.cs [hunk 12]
```diff
@@ -353,27 +359,75 @@ private async Task<SymmetricCryptoKey> MakePreloginKeyAsync(string masterPasswor
...
+
+                    var encKey = await _cryptoService.MakeEncKeyAsync(k);
+                    await _cryptoService.SetEncKeyAsync(encKey.Item2.EncryptedString);
+                    var keyPair = await _cryptoService.MakeKeyPairAsync();
+
+                    try
+                    {
+                        await _apiService.PostUserKeyToKeyConnector(tokenResponse.KeyConnectorUrl, keyConnectorRequest);
+                    }
+                    catch (ApiException e)
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Core/Services/AuthService.cs:415]: Also curious if there's a reason for the re-throws. `throw e` will invalidate the existing stack trace, so we might want to use `throw` instead. https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2200

File: src/Core/Models/Response/ProfileOrganizationResponse.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Core/Models/Response/ProfileOrganizationResponse.cs:28]: @eliykat suggestion seems reasonable, that's what we do in `jslib`.

```typescript
this.usesKeyConnector = this.getResponseProperty('UsesKeyConnector') ?? false;
```

File: src/App/Pages/TabsPage.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Hinton) [reply src/App/Pages/TabsPage.cs]: I think we need to check with @mportune-bw, I don't really have a better idea.

[the remark this answers is not in the document]
Comment (mportune-bw) [reply src/App/Pages/TabsPage.cs]: The core issue is that sync in mobile is performed asynchronously so the app is still usable while the vault is updated.  Unfortunately that's becoming problematic as more non-vault data is added to the sync model.

@vincentsalucci and I recently ran into this with the [Reset Master Password](https://github.com/bitwarden/mobile/pull/1492) flow, and eventually added the flag to the auth response (where it belongs imo).  Otherwise the experience becomes a clumsy mess depending on the size of the vault and network conditions (i.e. random screens popping up while the user is already logged in, and no guarantee the app will even be in the foreground when it happens, etc.)

@cscharf @kspearrin I believe a larger discussion is warranted as the "put everything in sync" approach is forcing mobile to make bad design decisions.

File: src/Core/Services/AuthService.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/Core/Services/AuthService.cs]: e6bfd661a20644aaf8c60347c08d5941e05b1a82

File: src/App/Pages/Settings/ExportVaultPage.xaml [hunk 13]
```diff
@@ -53,44 +54,54 @@
                 </StackLayout>
                 <Grid StyleClass="box-row">
                     <Grid.RowDefinitions>
+                        <RowDefinition Height="Auto" />
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/Settings/ExportVaultPage.xaml:57]: Xamarin defines multiple rows this way. Essentially builds a grid.

File: src/App/Pages/Accounts/RemoveMasterPasswordPageViewModel.cs [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+            _keyConnectorService = ServiceContainer.Resolve<IKeyConnectorService>("keyConnectorService");
+            _deviceActionService = ServiceContainer.Resolve<IDeviceActionService>("deviceActionService");
+            _apiService = ServiceContainer.Resolve<IApiService>("apiService");
+            _syncService = ServiceContainer.Resolve<ISyncService>("syncService");
+
+        }
+
+        public async Task Init()
+        {
+            this.Organization = await _keyConnectorService.GetManagingOrganization();
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/Accounts/RemoveMasterPasswordPageViewModel.cs]: 1e7bda3

File: src/Core/Models/Response/ProfileOrganizationResponse.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/Core/Models/Response/ProfileOrganizationResponse.cs:28]: b2d3eaa

File: src/App/Pages/Accounts/LockPage.xaml.cs [hunk 15]
```diff
@@ -130,7 +132,18 @@ private async Task UnlockedAsync()
                 return;
             }
             var previousPage = await AppHelpers.ClearPreviousPage();
-            Application.Current.MainPage = new TabsPage(_appOptions, previousPage);
+
+            if (await _vm.UserNeedsMigration())
+            {
+                var removeMasterPasswordPage = new RemoveMasterPasswordPage();
+                removeMasterPasswordPage.NavigateAction += () => Device.BeginInvokeOnMainThread(async () => await UnlockedAsync());
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/Accounts/LockPage.xaml.cs:139]: This adds an action to the RemoveMasterPasswordPage for navigating into the vault after the modal has been popped.
Will be removed if we change to `App.xaml.cs` and messaging service due to: https://github.com/bitwarden/mobile/pull/1633#discussion_r745282284

File: src/App/Pages/TabsPage.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (mportune-bw) [reply src/App/Pages/TabsPage.cs]: @jlf0dev Sounds like this prompt isn't time-sensitive so using `OnAppearing` on the TabsPage is fine for now.

File: src/App/Pages/Accounts/LockPage.xaml.cs [hunk 16]
```diff
@@ -12,6 +12,7 @@ namespace Bit.App.Pages
     public partial class LockPage : BaseContentPage
     {
         private readonly IStorageService _storageService;
+        private readonly IMessagingService _messagingService;
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/Accounts/LockPage.xaml.cs]: ee02697

File: src/App/Pages/TabsPage.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/TabsPage.cs:80]: I've moved the call for RemoveMasterPasswordPage to the `OnAppearing` of the TabsPage

ee02697

**Comment by jlf0dev** [comment]
Ready for final review @mportune-bw @eliykat @Hinton

File: src/App/App.xaml.cs [hunk 17]
```diff
@@ -140,6 +140,15 @@ public App(AppOptions appOptions)
                         }
                     });
                 }
+                else if (message.Command == "removeMasterPassword")
+                {
+                    Device.BeginInvokeOnMainThread(async () =>
+                    {
+                        await Application.Current.MainPage.Navigation.PushModalAsync(
+                        new NavigationPage(new RemoveMasterPasswordPage()));
```
Comment (eliykat) [inline src/App/App.xaml.cs:148]: Indentation

File: src/App/App.xaml.cs [hunk 18]
```diff
@@ -140,6 +140,15 @@ public App(AppOptions appOptions)
                         }
                     });
                 }
+                else if (message.Command == "removeMasterPassword")
```
Comment (eliykat) [inline src/App/App.xaml.cs:143]: Message command should be consistent with jslib (`convertAccountToKeyConnector`)

File: src/Core/Models/Response/ProfileOrganizationResponse.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply src/Core/Models/Response/ProfileOrganizationResponse.cs:28]: Strings are already nullable so I don't think it needs a default assignment. I was thinking of the `UsesKeyConnector` bool, should `= false` by default.

File: src/Core/Models/Response/IdentityTokenResponse.cs [hunk 19]
```diff
@@ -21,5 +21,6 @@ public class IdentityTokenResponse
         public KdfType Kdf { get; set; }
         public int? KdfIterations { get; set; }
         public bool ForcePasswordReset { get; set; }
+        public string KeyConnectorUrl { get; set; } = null;
```
Comment (eliykat) [inline src/Core/Models/Response/IdentityTokenResponse.cs:24]: I don't think you need the default assignment here, see my other comment in our thread below

File: src/App/Pages/TabsPage.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply src/App/Pages/TabsPage.cs:80]: So now we do the check on `TabsPage.OnAppearing` and send the message to `App.xaml.cs`, which then creates the new page. Seems fine

**Review by eliykat — CHANGES_REQUESTED** [review]
Final comments. Someone else should also review as I have code in here and shouldn't be the only reviewer. (@mportune-bw I'd appreciate you having a look over it)

File: src/App/App.xaml.cs (hunk 18 above)
Comment (mportune-bw) [reply src/App/App.xaml.cs:143]: still valid?

**Comment by mportune-bw** [comment]
Looks like @eliykat covered everything.  The layout on the export page could use some polish (is that screenshot up to date?)

**Comment by eliykat** [comment]
@jlf0dev added some spacing, it's not perfect but it looks a little nicer:

Screen Shot 2021-11-11 at 9 23 29 am

**Comment by mportune-bw** [comment]
> @jlf0dev added some spacing, it's not perfect but it looks a little nicer:

Yeah that's better.  The margin on the right side of the `Request` button is larger for some reason.

**Comment by jlf0dev** [comment]
> > @jlf0dev added some spacing, it's not perfect but it looks a little nicer:
> 
> 
> 
> Yeah that's better.  The margin on the right side of the `Request` button is larger for some reason.

Ah it probably needs `columnspan="2"`

File: src/Core/Models/Response/ProfileOrganizationResponse.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/Core/Models/Response/ProfileOrganizationResponse.cs:28]: Don't bools default to `false` anyway?

**Comment by jlf0dev** [comment]
image

File: src/Core/Models/Response/IdentityTokenResponse.cs (hunk 19 above)
Comment (jlf0dev) [reply src/Core/Models/Response/IdentityTokenResponse.cs]: removed null assignment from strings
68267be

File: src/App/App.xaml.cs (hunk 18 above)
Comment (jlf0dev) [reply src/App/App.xaml.cs]: 68267be

File: src/App/App.xaml.cs (hunk 17 above)
Comment (jlf0dev) [reply src/App/App.xaml.cs]: 68267be

File: src/Core/Models/Response/ProfileOrganizationResponse.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (eliykat) [reply src/Core/Models/Response/ProfileOrganizationResponse.cs:28]: Yes, apparently they do, all good

**Review by eliykat — APPROVED** [review]
Looks good to me; both Portune and Oscar have also looked over this at some point, everyone seems happy, so let's get on with it. Thanks again for your work on this!
